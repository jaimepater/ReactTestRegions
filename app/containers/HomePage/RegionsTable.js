import React from 'react';
import PropTypes from 'prop-types';
import { Table, Thead, Tr, Tbody, Td, H3 } from '@bootstrap-styled/v4';

const Row = ({ code, description, parentDescription, parentCode }) => (
  <Tr>
    <Td> {code} </Td>
    <Td>{description} </Td>
    <Td>{parentCode} </Td>
    <Td>{parentDescription} </Td>
  </Tr>
);

Row.propTypes = {
  code: PropTypes.string,
  description: PropTypes.string,
  parentDescription: PropTypes.string,
  parentCode: PropTypes.string,
};

const RegionsTable = ({ data, caption }) => (
  <Table striped>
    <caption>
      <H3>{caption}</H3>
    </caption>
    <Thead>
      <Tr>
        <Td>Codigo</Td>
        <Td>Descripcion</Td>
        <Td>Codigo padre</Td>
        <Td>Descripcionpadre</Td>
      </Tr>
    </Thead>
    <Tbody>
      {data.map((row, id) => (
        <Row key={id} {...row} />
      ))}
    </Tbody>
  </Table>
);

RegionsTable.propTypes = {
  data: PropTypes.array,
  caption: PropTypes.string,
};

export default RegionsTable;
