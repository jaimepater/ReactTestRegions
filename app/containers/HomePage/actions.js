/**
 * Set a regions reader for a file
 *
 * @param  {array}  regions The repository data
 * @return {object}   An action object with a type of SET_REGIONS
 */
export function setFiles(regions) {
  return {
    type: 'SET_REGIONS',
    regions,
  };
}
