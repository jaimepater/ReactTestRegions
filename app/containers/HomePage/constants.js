export const DELIMITATIONS = {
  columnDelimiter: ' / ',
  codeDelimiter: /(\d+)/,
};

export const LEVELS = {
  dep: {
    level: 0,
    label: 'Departamento',
  },
  pro: {
    level: 1,
    label: 'Provincia',
  },
  dis: {
    level: 2,
    label: 'Distrito',
  },
};
