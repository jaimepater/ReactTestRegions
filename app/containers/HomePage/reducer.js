const regionsReducer = function regions(state = [], action) {
  switch (action.type) {
    case 'SET_REGIONS':
      return action.regions;
    default:
      return state;
  }
};

export default regionsReducer;
