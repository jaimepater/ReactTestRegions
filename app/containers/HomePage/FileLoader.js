import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Form, FormGroup, H1, Input } from '@bootstrap-styled/v4';

import { DELIMITATIONS } from './constants';
import { convertTextToArray } from '../../utils/dataConvert';

class FileLoader extends Component {
  fileRead;

  handleFileRead = () => {
    const ArrayValues = convertTextToArray(
      this.fileRead.result,
      DELIMITATIONS.columnDelimiter,
      this.extractCodeName,
    );
    this.props.onSetFiles(ArrayValues);
  };

  handleFileSelector = file => {
    this.fileRead = new FileReader();
    this.fileRead.onloadend = this.handleFileRead;
    if (file instanceof Blob) {
      this.fileRead.readAsText(file);
    }
  };

  extractCodeName = row =>
    row
      .map(cell => {
        const infoRegion = cell.split(DELIMITATIONS.codeDelimiter);
        return infoRegion.length === 3
          ? { code: infoRegion[1], description: infoRegion[2].trim() }
          : null;
      })
      .filter(line => line != null);

  render() {
    return (
      <div className="mt-3">
        <H1>Tabla por Regiones</H1>
        <Form>
          <FormGroup>
            <Input
              type="file"
              id="file"
              size="sm"
              onChange={e => this.handleFileSelector(e.target.files[0])}
            />
          </FormGroup>
        </Form>
      </div>
    );
  }
}

FileLoader.propTypes = {
  onSetFiles: PropTypes.func,
};

export default FileLoader;
