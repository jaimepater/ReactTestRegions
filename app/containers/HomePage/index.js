/*
 * HomePage
 *
 * This is the first thing users see of our App, at the '/' route
 *
 */

import React from 'react';
import { compose } from 'redux';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { useInjectReducer } from 'utils/injectReducer';
import { Container } from '@bootstrap-styled/v4';

import FileLoader from './FileLoader';
import RegionsTable from './RegionsTable';
import { LEVELS } from './constants';
import { setFiles } from './actions';
import reducer from './reducer';

const key = 'regions';

function mapStateToProps(state) {
  return {
    home: state,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    onSetFiles: evt => dispatch(setFiles(evt)),
  };
}

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

function HomePage(props) {
  console.log(props);
  useInjectReducer({ key, reducer });

  const getInformationForLevel = level => {
    const data = props.home.regions;
    return data
      .filter(row => row.length === level + 1 && row[level])
      .map(row => {
        const parentValue =
          level - 1 < 0 ? { code: '-', description: '-' } : row[level - 1];
        return {
          ...row[level],
          parentCode: parentValue.code,
          parentDescription: parentValue.description,
        };
      });
  };
  return (
    <section>
      <Container>
        <FileLoader {...props} />
        {props.home.regions && props.home.regions.length ? (
          <div>
            <RegionsTable
              data={getInformationForLevel(LEVELS.dep.level)}
              caption={LEVELS.dep.label}
            />
            <RegionsTable
              data={getInformationForLevel(LEVELS.pro.level)}
              caption={LEVELS.pro.label}
            />
            <RegionsTable
              data={getInformationForLevel(LEVELS.dis.level)}
              caption={LEVELS.dis.label}
            />
          </div>
        ) : (
          <div />
        )}
      </Container>
    </section>
  );
}

HomePage.propTypes = {
  home: PropTypes.object,
};

export default compose(withConnect)(HomePage);
