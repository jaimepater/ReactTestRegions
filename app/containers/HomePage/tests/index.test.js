import React from 'react';
import { render } from 'react-testing-library';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import { browserHistory } from 'react-router-dom';
import HomePage from '../index';
import configureStore from '../../../configureStore';

describe('<HomePage />', () => {
  let store;

  beforeAll(() => {
    store = configureStore({}, browserHistory);
  });
  it('should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <HomePage loading={false} error={false} repos={[]} />
        </IntlProvider>
      </Provider>,
    );
    expect(firstChild).toMatchSnapshot();
  });
  describe('getInformationForLevel', () => {
    it('should dispatch changeLocale when called', () => {
      const dispatch = jest.fn();
      const result = HomePage.getInformationForLevel(2);
      expect(result).toBe(3);
    });
  });
});
