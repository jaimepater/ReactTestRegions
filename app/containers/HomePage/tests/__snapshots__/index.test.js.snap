// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> should render and match the snapshot 1`] = `
<section>
  <div
    class="Container-u38a83-0 jDuhNh container"
  >
    <div
      class="mt-3"
    >
      <h1
        class="H1-xsrgru-0 dYBbvb"
      >
        Tabla por Regiones
      </h1>
      <form
        class="Form-sc-15bnz0c-0 jEhIrw"
      >
        <div
          class="form-group"
        >
          <input
            class="Input-wun2o8-0 bmUDTN form-control-sm form-control-file"
            id="file"
            type="file"
          />
        </div>
      </form>
    </div>
    <div />
  </div>
</section>
`;
